File: libs/OscFramework_2.0/Source Code/Samples/OscDemo/VB/Transmitter/TcpTransmitter.vb
Public Class TcpTransmitter
    Implements ITransmitter

    Public Sub Start(packet As OscPacket) Implements ITransmitter.Start
        Assert.ParamIsNotNull(packet)

        mOscClient = New OscClient(Destination)
        mOscClient.Connect()

        mPacket = packet
        mPacket.Client = mOscClient
        Thread.VolatileWrite(mSendMessages, True)

        mTransmitterThread = New Thread(AddressOf RunWorker)
        mTransmitterThread.Start()
    End Sub

    Public Sub [Stop]() Implements ITransmitter.Stop
        Thread.VolatileWrite(mSendMessages, False)
        mTransmitterThread.Join()

        mOscClient.Close()
    End Sub

    Private Sub RunWorker()
        Try
            While Thread.VolatileRead(mSendMessages)
                mPacket.Send()

                mTransmissionCount += 1
                Console.Clear()
                Console.WriteLine("Osc Transmitter: Tcp")
                Console.WriteLine("Transmission Count: {0}{1}", mTransmissionCount, Environment.NewLine)
                Console.WriteLine("Press any key to exit.")

                Thread.Sleep(1000)
            End While
        Catch ex As Exception
            Console.WriteLine(ex.Message)
        End Try
    End Sub

    Private ReadOnly Destination As IPEndPoint = New IPEndPoint(IPAddress.Loopback, Program.Port)

    Private mSendMessages As Boolean
    Private mTransmitterThread As Thread
    Private mPacket As OscPacket
    Private mOscClient As OscClient
    Private mTransmissionCount As Integer
End Class

File: libs/OscFramework_2.0/Source Code/Samples/OscDemo/CS/Transmitter/TcpTransmitter.cs
using System;
using System.Threading;
using System.Net;
using Bespoke.Common;
using Bespoke.Common.Osc;

namespace Transmitter
{
    public class TcpTransmitter : ITransmitter
    {
        public void Start(OscPacket packet)
        {
            Assert.ParamIsNotNull(packet);

            mOscClient = new OscClient(Destination);
            mOscClient.Connect();

            mPacket = packet;
            mPacket.Client = mOscClient;
            mSendMessages = true;

            mTransmitterThread = new Thread(RunWorker);
            mTransmitterThread.Start();
        }

        public void Stop()
        {
            mSendMessages = false;
            mTransmitterThread.Join();

            mOscClient.Close();
        }

        private void RunWorker()
        {
            try
            {
                while (mSendMessages)
                {
                    mPacket.Send();

                    mTransmissionCount++;
                    Console.Clear();
                    Console.WriteLine("Osc Transmitter: Tcp");
                    Console.WriteLine("Transmission Count: {0}\n", mTransmissionCount);
                    Console.WriteLine("Press any key to exit.");

                    Thread.Sleep(1000);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        private static readonly IPEndPoint Destination = new IPEndPoint(IPAddress.Loopback, Program.Port);

        private volatile bool mSendMessages;
        private Thread mTransmitterThread;
        private OscPacket mPacket;
        private OscClient mOscClient;
        private int mTransmissionCount;        
    }
}
